src/dfx/src/commands/canister/call.rs

```diff
@@ -209,29 +212,34 @@ pub async fn exec(
     opts: CanisterCallOpts,
     call_sender: &CallSender,
 ) -> DfxResult {
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+    fetch_root_key_if_needed(env).await?;
+
     let callee_canister = opts.canister_name.as_str();
     let method_name = opts.method_name.as_str();
     let canister_id_store = CanisterIdStore::for_env(env)?;
 
-    let (canister_id, maybe_candid_path) = match CanisterId::from_text(callee_canister) {
-        Ok(id) => {
-            if let Some(canister_name) = canister_id_store.get_name(callee_canister) {
-                get_local_cid_and_candid_path(env, canister_name, Some(id))?
-            } else {
-                // TODO fetch candid file from remote canister
-                (id, None)
-            }
-        }
+    let canister_id = match CanisterId::from_text(callee_canister) {
+        Ok(id) => id,
         Err(_) => {
             let canister_id = canister_id_store.get(callee_canister)?;
-            get_local_cid_and_candid_path(env, callee_canister, Some(canister_id))?
+            get_local_cid_and_candid_path(env, callee_canister, Some(canister_id))?.0
         }
     };
-    let maybe_candid_path = opts.candid.or(maybe_candid_path);
+    let method_type = if let Some(path) = opts.candid {
+        get_candid_type(CandidSource::Path(&path), method_name)
+    } else {
+        read_module_metadata(agent, canister_id, "candid:service")
```

Comment:
The idea is we will never use the local candid file?  What if the canister does not have candid:service metadata, or if it's not public?  I would have expected this to fall back to reading the candid:service metadata if the local candid file was not found, not to skip it entirely.

Reply:
Yes, the assumption is that `dfx build` or other build tool have already put the local did file into the metadata. Unless you are the author of the canister, you won't have the local did file anyway, so reading the local did file is redundant. 

If the metadata is not there, dfx emits a warning, and hopefully the developer can fix that. If we read the local did file first, the canister owner will have less incentive to add metadata. So this is also a gentle push into a better ecosystem.

Reply:
I discussed this with @sesi200. The conclusion of our discussion is that reading the candid:service metadata is a good approach to ensure that the current version is used. For the sake of backward compatibility, the local version should not be ignored, though. We feel that the following flow would make sense:

If a path is provided on the command line, the path is used to read the candid information.
If no path is provided, `dfx` tries to get the candid information from the metadata. If it succeeds to retrieve this information, it is used. If the information could not be retrieved, `dfx` tries to use the local `dfx.json` to get the candid information.

Does that make sense to you?

Reply:
@THLO This PR implements the workflow you mentioned except for the last item.

> If the information could not be retrieved, dfx tries to use the local dfx.json to get the candid information.

It's a tech debt from the beginning due to the lack of metadata in the canister. We should aim to remove this reliance, instead of maintaining the tech debt. This change also won't break one's workflow unless they didn't put the candid info as metadata. In which case, we would like to nudge them to do so, instead of silently fallback.

Conceptually, everything under `.dfx` is a build artifact and they shouldn't be used when calling canisters. This ensures that the canister owner and other users have the same experience. A build process can fail, and you may end up with an out-of-sync local did file. We don't want to rely on the build artifact to call canisters. Happy to schedule a meeting to discuss this further.

Reply:
I see your point. @sesi200 convinced me that maintaining backward compatibility would be nice but maybe it's not appropriate here. I think a quick call would be helpful.